From a3f29c8002d36171018b87770957e11203028e5d Mon Sep 17 00:00:00 2001
From: Simeon Marijon <simeon.marijon@foss.st.com>
Date: Thu, 7 Dec 2023 14:47:23 +0100
Subject: [PATCH 435/954] feat(st): add support for TAMP backup registers

For both STM32MP1 and STM32MP2 it allows access to the
backup registers checking access right based on the zones
described in the device tree.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b296fe5a57c2e6eb0fa2cb4481c56005d3e7afb3 feat(st): add support for TAMP backup registers]
Signed-off-by: Simeon Marijon <simeon.marijon@foss.st.com>
Change-Id: Ie6a60a7d9c5ecbb5faa605ae19fb9cb97fe7deb5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/343949
Tested-by: Simeon MARIJON <simeon.marijon@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Simeon MARIJON <simeon.marijon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/st/nvmem/stm32mp_tamp_nvram_core.c | 417 +++++++++++++++++++++
 drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c  |  43 +++
 drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c  | 232 ++++++++++++
 include/drivers/st/stm32mp_tamp_nvram.h    |  83 ++++
 plat/st/common/common.mk                   |   2 +
 plat/st/common/include/stm32mp_common.h    |   3 +
 plat/st/stm32mp1/bl2_plat_setup.c          |   4 +
 plat/st/stm32mp1/platform.mk               |   2 +
 plat/st/stm32mp1/stm32mp1_def.h            |   1 +
 plat/st/stm32mp2/bl2_plat_setup.c          |   4 +
 plat/st/stm32mp2/platform.mk               |   2 +
 plat/st/stm32mp2/stm32mp2_def.h            |   1 +
 12 files changed, 794 insertions(+)
 create mode 100644 drivers/st/nvmem/stm32mp_tamp_nvram_core.c
 create mode 100644 drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
 create mode 100644 drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c
 create mode 100644 include/drivers/st/stm32mp_tamp_nvram.h

diff --git a/drivers/st/nvmem/stm32mp_tamp_nvram_core.c b/drivers/st/nvmem/stm32mp_tamp_nvram_core.c
new file mode 100644
index 0000000000..e12d1d2490
--- /dev/null
+++ b/drivers/st/nvmem/stm32mp_tamp_nvram_core.c
@@ -0,0 +1,417 @@
+/*
+ * Copyright (c) 2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+
+#include <drivers/clk.h>
+#include <drivers/delay_timer.h>
+#include <drivers/st/nvmem.h>
+#include <drivers/st/stm32mp_tamp_nvram.h>
+#include <lib/mmio.h>
+#include <libfdt.h>
+
+#include <platform_def.h>
+#include <stm32mp_dt.h>
+
+#define DEBUG_RIGHT_MAX_SIZE 100
+
+static struct stm32_tamp_nvram_drv_data nvram_drv_data;
+
+int stm32_tamp_nvram_reg_get_zone(struct stm32_tamp_nvram_drv_data *drv_data, int reg)
+{
+	int *bkpreg_zones_end = drv_data->bkpreg_zones_end;
+	unsigned int nb_zones = drv_data->cdata->nb_zones;
+	unsigned int protection_zone_idx;
+
+	for (protection_zone_idx = 0; protection_zone_idx < nb_zones; protection_zone_idx++) {
+		if (reg <= bkpreg_zones_end[protection_zone_idx]) {
+			break;
+		}
+	}
+
+	if (protection_zone_idx >= nb_zones) {
+		return -1; /* the reg is not a part of any zone */
+	}
+
+	return protection_zone_idx;
+}
+
+static int
+stm32_tamp_nvram_read_bkp(struct stm32_tamp_nvram_drv_data *drv_data, size_t offset, uint32_t *val)
+{
+	unsigned int reg = (offset / sizeof(uint32_t));
+	int protection_zone_idx = stm32_tamp_nvram_reg_get_zone(drv_data, (int)reg);
+	int ret = -1;
+
+	if (protection_zone_idx >= 0) {
+		bool clk_should_be_toggled = !clk_is_enabled(TAMP_BKP_REG_CLK);
+
+		switch (drv_data->bkpreg_access[protection_zone_idx]) {
+		case RNSWNS:
+		case RNSWS:
+		case RONS:
+		case RSWS:
+			if (clk_should_be_toggled) {
+				clk_enable(TAMP_BKP_REG_CLK);
+			}
+			*val = mmio_read_32(drv_data->base + offset);
+			if (clk_should_be_toggled) {
+				clk_disable(TAMP_BKP_REG_CLK);
+			}
+			ret = 0;
+			break;
+		case NORNOW:
+			*val = U(0x0);
+			ret = 0;
+			break;
+		default:
+			ERROR("Wrong access rights for zone %d (%u)\n", protection_zone_idx, reg);
+		}
+	}
+
+	return ret;
+}
+
+static int
+stm32_tamp_nvram_write_bkp(struct stm32_tamp_nvram_drv_data *drv_data, size_t offset, uint32_t val)
+{
+	unsigned int reg = (offset / sizeof(uint32_t));
+	int protection_zone_idx = stm32_tamp_nvram_reg_get_zone(drv_data, (int)reg);
+	int ret = -1;
+
+	if (protection_zone_idx >= 0) {
+		bool clk_should_be_toggled = !clk_is_enabled(TAMP_BKP_REG_CLK);
+
+		switch (drv_data->bkpreg_access[protection_zone_idx]) {
+		case RNSWNS:
+		case RNSWS:
+		case RSWS:
+			if (clk_should_be_toggled) {
+				clk_enable(TAMP_BKP_REG_CLK);
+			}
+			mmio_write_32(drv_data->base + offset, val);
+			if (clk_should_be_toggled) {
+				clk_disable(TAMP_BKP_REG_CLK);
+			}
+			ret = 0;
+			break;
+		case RONS:
+		case NORNOW:
+			ret = 0;
+			break;
+		default:
+			ERROR("Wrong access rights for zone %d (%u)\n", protection_zone_idx, reg);
+		}
+	}
+
+	return ret;
+}
+
+int stm32_tamp_nvram_reg_field_read(uintptr_t base, struct stm32_tamp_nvram_reg_field field,
+				    uint32_t *val)
+{
+	uint32_t reg_val = 0;
+
+	reg_val = mmio_read_32(base + field.reg_offset);
+	reg_val &= GENMASK_32(field.highest_bit, field.lowest_bit);
+	reg_val >>= field.lowest_bit;
+	*val = reg_val;
+
+	return 0;
+}
+
+int stm32_tamp_nvram_reg_field_write(uintptr_t base, struct stm32_tamp_nvram_reg_field field,
+				     uint32_t val)
+{
+	uint32_t reg_val = 0;
+	uint32_t val_temp = val;
+	bool clk_should_be_toggled = !clk_is_enabled(TAMP_BKP_REG_CLK);
+
+	val_temp &= GENMASK_32(field.highest_bit - field.lowest_bit, 0);
+	if (clk_should_be_toggled) {
+		clk_enable(TAMP_BKP_REG_CLK);
+	}
+	reg_val = mmio_read_32(base + field.reg_offset);
+	reg_val &= ~(GENMASK_32(field.highest_bit, field.lowest_bit));
+	reg_val |= (val_temp << field.lowest_bit);
+	mmio_write_32(base + field.reg_offset, reg_val);
+	if (clk_should_be_toggled) {
+		clk_disable(TAMP_BKP_REG_CLK);
+	}
+
+	return 0;
+}
+
+static int
+stm32_tamp_nvram_write_byte(struct stm32_tamp_nvram_drv_data *drv_data, size_t offset, uint8_t byte)
+{
+	size_t offset_aligned = round_down(offset, sizeof(uint32_t));
+	size_t byte_in_word = offset - offset_aligned;
+	uint32_t read_value = 0;
+	uint32_t to_be_writen_value = 0;
+	uint32_t mask = UCHAR_MAX;
+
+	mask = ~(mask << (byte_in_word * U(8)));
+
+	stm32_tamp_nvram_read_bkp(drv_data, offset_aligned, &read_value);
+
+	to_be_writen_value = read_value & mask;
+	to_be_writen_value |= (uint32_t)byte << (byte_in_word * U(8));
+
+	return stm32_tamp_nvram_write_bkp(drv_data, offset_aligned, to_be_writen_value);
+}
+
+static int
+stm32_tamp_nvram_read_byte(struct stm32_tamp_nvram_drv_data *drv_data, size_t offset, uint8_t *byte)
+{
+	size_t offset_aligned = round_down(offset, sizeof(uint32_t));
+	size_t byte_in_word = offset - offset_aligned;
+	uint32_t read_value = 0;
+
+	stm32_tamp_nvram_read_bkp(drv_data, offset_aligned, &read_value);
+	*byte = (uint8_t)((read_value >> (byte_in_word * U(8))) & UCHAR_MAX);
+
+	return 0;
+}
+
+static int
+stm32_tamp_nvram_read(struct stm32_tamp_nvram_drv_data *drv_data, size_t offset, uint8_t *buf_u8,
+		      size_t size)
+{
+	size_t i = offset;
+	size_t total = offset + size;
+
+	while (i < total) {
+		size_t reg_idx = i / sizeof(uint32_t);
+
+		if (((i + sizeof(uint32_t)) <= total) && is_aligned(i, sizeof(uint32_t))) {
+			uint32_t tmp_u32 = 0;
+			void *mem_cpy_ret = NULL;
+
+			stm32_tamp_nvram_read_bkp(drv_data, i, &tmp_u32);
+			mem_cpy_ret = memcpy(&buf_u8[i - offset],
+					     (uint8_t *)&tmp_u32,
+					     sizeof(uint32_t));
+			if (mem_cpy_ret != &buf_u8[i - offset]) {
+				ERROR("Failed to read bkp %zu\n", reg_idx);
+
+				return -1;
+			}
+			i += sizeof(uint32_t);
+		} else {
+			uint8_t byte = 0;
+			int ret = stm32_tamp_nvram_read_byte(drv_data, i, &byte);
+
+			if (ret != 0) {
+				ERROR("BKPreg %zu is not allowed to be read\n", reg_idx);
+				byte = 0;
+			}
+			buf_u8[i] = byte;
+			i++;
+		}
+	}
+
+	return 0;
+}
+
+static int
+stm32_tamp_nvram_write(struct stm32_tamp_nvram_drv_data *drv_data, size_t offset,
+		       const uint8_t *buf_u8, size_t size)
+{
+	size_t total = offset + size;
+	size_t i = offset;
+
+	while (i < total) {
+		int ret = 0;
+		size_t reg_idx = i / sizeof(uint32_t);
+
+		if (((i + sizeof(uint32_t)) <= total) && is_aligned(i, sizeof(uint32_t))) {
+			uint32_t tmp_u32 = 0;
+			void *mem_cpy_ret = NULL;
+
+			mem_cpy_ret = memcpy((uint8_t *)&tmp_u32,
+					     &buf_u8[i - offset],
+					     sizeof(uint32_t));
+			if (mem_cpy_ret != &tmp_u32) {
+				ERROR("Failed to read bkp %zu\n", reg_idx);
+				return -1;
+			}
+
+			ret = stm32_tamp_nvram_write_bkp(drv_data, i, tmp_u32);
+			if (ret != 0) {
+				ERROR("Failed to read bkp %zu\n", reg_idx);
+				return ret;
+			}
+			i += sizeof(uint32_t);
+		} else {
+			ret = stm32_tamp_nvram_write_byte(drv_data, i, buf_u8[i - offset]);
+			if (ret != 0) {
+				VERBOSE("BKPreg %zu is not allowed to be written\n", reg_idx);
+			}
+			i++;
+		}
+	}
+
+	return 0;
+}
+
+static int *
+stm32_tamp_nvram_get_backup_zones(struct stm32_tamp_nvram_drv_data *drv_data)
+{
+	static int static_bkpreg_zones_end[MAX_TAMP_BACKUP_REGS_ZONES];
+	unsigned int nb_zones = drv_data->cdata->nb_zones;
+	const struct stm32_tamp_nvram_cdata *cdata = drv_data->cdata;
+	unsigned int zone_idx = 0;
+	int *bkpreg_zones_end = static_bkpreg_zones_end;
+	uint32_t offset_field = 0;
+	struct stm32_tamp_nvram_reg_field zone_cfg_field = { 0 };
+
+	/* Get the n-1 frontiers of zone within the tamp configuration registers */
+	for (zone_idx = 0; zone_idx < (nb_zones - U(1)); zone_idx++) {
+		zone_cfg_field = cdata->zone_cfg[zone_idx];
+
+		if (stm32_tamp_nvram_reg_field_read(drv_data->parent_base,
+						    zone_cfg_field,
+						    &offset_field) != 0) {
+			ERROR("Can't read field for registers zones\n");
+			return NULL;
+		}
+
+		/* in order to deal with no offset for a zone we store -1 for the zone end
+		 * => the zone is empty
+		 */
+		bkpreg_zones_end[zone_idx] = ((int)offset_field) - 1;
+	}
+
+	/* The last zone end is defined by the number of registers in TAMP */
+	bkpreg_zones_end[zone_idx] = ((int)drv_data->nb_total_regs) - 1;
+
+	return bkpreg_zones_end;
+}
+
+static int
+stm32_tamp_nvram_read_cell(struct nvmem_cell *cell, uint8_t *output_buf, size_t buf_len,
+				      size_t *read_len)
+{
+	struct stm32_tamp_nvram_drv_data *drv_data =
+		(struct stm32_tamp_nvram_drv_data *)cell->dev->drv_data;
+
+	if (buf_len < cell->len) {
+		return -1;
+	}
+
+	if (read_len != NULL) {
+		*read_len = cell->len;
+	}
+
+	return stm32_tamp_nvram_read(drv_data, cell->offset, output_buf, cell->len);
+}
+
+static int stm32_tamp_nvram_write_cell(struct nvmem_cell *cell, uint8_t *input_buf, size_t len)
+{
+	struct stm32_tamp_nvram_drv_data *drv_data =
+		(struct stm32_tamp_nvram_drv_data *)cell->dev->drv_data;
+
+	if (len > cell->len) {
+		return -1;
+	}
+
+	return stm32_tamp_nvram_write(drv_data, cell->offset, input_buf, cell->len);
+}
+
+static const struct nvmem_ops stm32_tamp_nvram_nvmem_ops = {
+	.read_cell = stm32_tamp_nvram_read_cell,
+	.write_cell = stm32_tamp_nvram_write_cell,
+};
+
+int stm32_tamp_nvram_init(void)
+{
+	int node = 0;
+	int parent_node = 0;
+	void *fdt = NULL;
+	int len = 0;
+	int ret;
+	const fdt32_t *reg_prop = NULL;
+	const fdt32_t *parent_reg_prop = NULL;
+
+	if (fdt_get_address(&fdt) == 0) {
+		return -ENODEV;
+	}
+
+	/* Get node and compatible data */
+	node = fdt_node_offset_by_compatible(fdt, -1, DT_TAMP_NVRAM_COMPAT);
+	if (node < 0) {
+		return -ENODEV;
+	}
+
+	nvram_drv_data.cdata = stm32_tamp_nvram_get_cdata();
+
+	reg_prop = fdt_getprop(fdt, node, "reg", &len);
+
+	if ((reg_prop == NULL) || ((unsigned int)len != (sizeof(uint32_t) * U(2)))) {
+		return -EINVAL;
+	}
+
+	nvram_drv_data.base = fdt32_to_cpu(reg_prop[0]);
+	nvram_drv_data.size = fdt32_to_cpu(reg_prop[1]);
+
+	parent_node = fdt_parent_offset(fdt, node);
+	if (parent_node < 0) {
+		return -EINVAL;
+	}
+
+	/* Get address of the device */
+	parent_reg_prop = fdt_getprop(fdt, parent_node, "reg", &len);
+
+	if ((parent_reg_prop == NULL) || ((unsigned int)len != (sizeof(uint32_t) * U(2)))) {
+		return -EINVAL;
+	}
+
+	nvram_drv_data.parent_base = fdt32_to_cpu(parent_reg_prop[0]);
+	nvram_drv_data.parent_size = fdt32_to_cpu(parent_reg_prop[1]);
+
+	nvram_drv_data.nb_total_regs = nvram_drv_data.size / sizeof(uint32_t);
+
+	ret = stm32_tamp_nvram_update_rights();
+	if (ret != 0) {
+		return ret;
+	}
+
+	nvram_drv_data.node = node;
+	nvram_drv_data.tamp_fdt = fdt;
+
+	nvmem_driver_register(nvram_drv_data.node, &stm32_tamp_nvram_nvmem_ops, &nvram_drv_data);
+
+	return 0;
+}
+
+int stm32_tamp_nvram_update_rights(void)
+{
+	bool clk_should_be_toggled = !clk_is_enabled(TAMP_BKP_REG_CLK);
+
+	if (clk_should_be_toggled) {
+		clk_enable(TAMP_BKP_REG_CLK);
+	}
+	nvram_drv_data.bkpreg_zones_end = stm32_tamp_nvram_get_backup_zones(&nvram_drv_data);
+	if (nvram_drv_data.bkpreg_zones_end == NULL) {
+		NOTICE("Failed to get the backup zone from tamp regs\n\n");
+		return -ENOMEM;
+	}
+
+	nvram_drv_data.bkpreg_access = stm32_tamp_nvram_get_access_rights(&nvram_drv_data);
+	if (nvram_drv_data.bkpreg_access == NULL) {
+		ERROR("Failed to get the backup zone access rights\n\n");
+		return -ENOMEM;
+	}
+
+	if (clk_should_be_toggled) {
+		clk_disable(TAMP_BKP_REG_CLK);
+	}
+
+	return 0;
+}
diff --git a/drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c b/drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
new file mode 100644
index 0000000000..5b0fa5a1f0
--- /dev/null
+++ b/drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <common/debug.h>
+#include <drivers/st/stm32mp_tamp_nvram.h>
+#include <lib/mmio.h>
+
+#include <platform_def.h>
+
+#define BKPREG_ZONE_1 U(0)
+#define BKPREG_ZONE_2 U(1)
+#define BKPREG_ZONE_3 U(2)
+
+static const struct stm32_tamp_nvram_reg_field
+	mp1_zone_cfg_fields[MAX_TAMP_BACKUP_REGS_ZONES - 1] = {
+		[BKPREG_ZONE_1] = REG_FIELD(TAMP_SECCFGR, 0, 7),
+		[BKPREG_ZONE_2] = REG_FIELD(TAMP_SECCFGR, 16, 23),
+	};
+
+static const struct stm32_tamp_nvram_cdata stm32mp1_tamp_nvram = {
+	.nb_zones = MAX_TAMP_BACKUP_REGS_ZONES,
+	.zone_cfg = mp1_zone_cfg_fields,
+};
+
+static enum stm32_tamp_bkpreg_access mp1_bkpreg_access[MAX_TAMP_BACKUP_REGS_ZONES] = {
+	[BKPREG_ZONE_1] = RSWS,
+	[BKPREG_ZONE_2] = RNSWS,
+	[BKPREG_ZONE_3] = RNSWNS,
+};
+
+const struct stm32_tamp_nvram_cdata *stm32_tamp_nvram_get_cdata(void)
+{
+	return &stm32mp1_tamp_nvram;
+}
+
+enum stm32_tamp_bkpreg_access *
+stm32_tamp_nvram_get_access_rights(struct stm32_tamp_nvram_drv_data *drv_data)
+{
+	return &mp1_bkpreg_access[0];
+}
diff --git a/drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c b/drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c
new file mode 100644
index 0000000000..a0df9c0ba3
--- /dev/null
+++ b/drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <common/debug.h>
+#include <drivers/st/stm32mp_tamp_nvram.h>
+#include <lib/mmio.h>
+
+#include <platform_def.h>
+
+#define BKPREG_ZONE_1_RIF1 U(0)
+#define BKPREG_ZONE_1_RIF2 U(1)
+#define BKPREG_ZONE_2_RIF1 U(2)
+#define BKPREG_ZONE_2_RIF2 U(3)
+#define BKPREG_ZONE_3_RIF1 U(4)
+#define BKPREG_ZONE_3_RIF0 U(5)
+#define BKPREG_ZONE_3_RIF2 U(6)
+#define NB_CMPT_MP2 U(3)
+
+static const struct stm32_tamp_nvram_reg_field
+	mp25_zone_cfg_fields[MAX_TAMP_BACKUP_REGS_ZONES - 1] = {
+		[BKPREG_ZONE_1_RIF1] = REG_FIELD(TAMP_BKPRIFR(1), U(0), U(7)),
+		[BKPREG_ZONE_1_RIF2] = REG_FIELD(TAMP_SECCFGR, U(0), U(7)),
+		[BKPREG_ZONE_2_RIF1] = REG_FIELD(TAMP_BKPRIFR(2), U(0), U(7)),
+		[BKPREG_ZONE_2_RIF2] = REG_FIELD(TAMP_SECCFGR, U(16), U(23)),
+		[BKPREG_ZONE_3_RIF1] = REG_FIELD(TAMP_BKPRIFR(3), U(0), U(7)),
+		[BKPREG_ZONE_3_RIF0] = REG_FIELD(TAMP_BKPRIFR(3), U(16), U(23)),
+	};
+
+static const struct stm32_tamp_nvram_reg_field
+	mp25_rxcidcfg_cfen_fields[NB_CMPT_MP2] = {
+		REG_FIELD(TAMP_RXCIDCFGR(0), U(0), U(0)),
+		REG_FIELD(TAMP_RXCIDCFGR(1), U(0), U(0)),
+		REG_FIELD(TAMP_RXCIDCFGR(2), U(0), U(0)),
+	};
+
+static const struct stm32_tamp_nvram_reg_field
+	mp25_rxcidcfg_fields[NB_CMPT_MP2] = {
+		REG_FIELD(TAMP_RXCIDCFGR(0), U(4), U(6)),
+		REG_FIELD(TAMP_RXCIDCFGR(1), U(4), U(6)),
+		REG_FIELD(TAMP_RXCIDCFGR(2), U(4), U(6)),
+	};
+
+static const struct stm32_tamp_nvram_cdata stm32mp25_tamp_nvram = {
+	.nb_zones = MAX_TAMP_BACKUP_REGS_ZONES,
+	.zone_cfg = mp25_zone_cfg_fields,
+};
+
+const struct stm32_tamp_nvram_cdata *stm32_tamp_nvram_get_cdata(void)
+{
+	return &stm32mp25_tamp_nvram;
+}
+
+static int stm32_tamp_nvram_is_isolation_en_mp2(
+	const struct stm32_tamp_nvram_drv_data *drv_data)
+{
+	unsigned int nb_compartment_enabled = 0;
+	uint32_t cfen = 0;
+	struct stm32_tamp_nvram_reg_field cfen_field = { 0 };
+	int ret = 1;
+
+	for (unsigned int i = 0; i < NB_CMPT_MP2; i++) {
+		cfen_field = mp25_rxcidcfg_cfen_fields[i];
+		if (stm32_tamp_nvram_reg_field_read(drv_data->parent_base,
+						    cfen_field, &cfen) != 0) {
+			NOTICE("TFA: Can't read field for registers zones\n");
+			nb_compartment_enabled = UINT_MAX;
+			break;
+		}
+
+		nb_compartment_enabled += cfen;
+	}
+
+	if (nb_compartment_enabled == U(0)) {
+		ret = 0;
+	} else if (nb_compartment_enabled == NB_CMPT_MP2) {
+		ret = 1;
+	} else {
+		ret = -1;
+	}
+
+	return ret;
+}
+
+static bool *
+stm32_tamp_get_cmpt_owner_mp2(struct stm32_tamp_nvram_drv_data *drv_data)
+{
+	struct stm32_tamp_nvram_reg_field cid_field = { U(0) };
+	uint32_t cid_per_zone = U(0);
+	int isolation_enabled = 0;
+	static bool compartment_owner[3];
+	unsigned int i = 0;
+	bool *ret = NULL;
+
+	isolation_enabled = stm32_tamp_nvram_is_isolation_en_mp2(drv_data);
+	if (isolation_enabled >= 0) {
+		ret = compartment_owner;
+		for (i = U(0); i < NB_CMPT_MP2; i++) {
+			if (isolation_enabled != 0) {
+				cid_field = mp25_rxcidcfg_fields[i];
+
+				if (stm32_tamp_nvram_reg_field_read(drv_data->parent_base,
+						   cid_field,
+						   &cid_per_zone) != 0) {
+					NOTICE("TFA: Can't read field for registers zones\n");
+					ret = NULL;
+					break;
+				}
+
+				if (cid_per_zone == CURRENT_CID) {
+					compartment_owner[i] = true;
+				} else {
+					compartment_owner[i] = false;
+				}
+			} else {
+				compartment_owner[i] = true;
+			}
+		}
+	}
+
+	return ret;
+}
+
+enum stm32_tamp_bkpreg_access *
+stm32_tamp_nvram_get_access_rights(struct stm32_tamp_nvram_drv_data *drv_data)
+{
+	static enum stm32_tamp_bkpreg_access static_bkpreg_access[MAX_TAMP_BACKUP_REGS_ZONES];
+	unsigned int nb_zones = drv_data->cdata->nb_zones;
+	enum stm32_tamp_bkpreg_access *bkpreg_access = static_bkpreg_access;
+
+	drv_data->compartment_owner = stm32_tamp_get_cmpt_owner_mp2(drv_data);
+	if (drv_data->compartment_owner != NULL) {
+		bkpreg_access = static_bkpreg_access;
+		for (unsigned int protection_zone_idx = 0;
+		     (protection_zone_idx < nb_zones); protection_zone_idx++) {
+			switch (protection_zone_idx) {
+			case BKPREG_ZONE_1_RIF1:
+				if (drv_data->compartment_owner[U(1)]) {
+					bkpreg_access[protection_zone_idx] =
+						RSWS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			case BKPREG_ZONE_1_RIF2:
+				if (drv_data->compartment_owner[U(2)]) {
+					bkpreg_access[protection_zone_idx] =
+						RSWS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			case BKPREG_ZONE_2_RIF1:
+				if (drv_data->compartment_owner[U(1)]) {
+					bkpreg_access[protection_zone_idx] =
+						RNSWS;
+				} else if (drv_data->compartment_owner[U(2)]) {
+					bkpreg_access[protection_zone_idx] =
+						RONS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			case BKPREG_ZONE_2_RIF2:
+				if (drv_data->compartment_owner[U(2)]) {
+					bkpreg_access[protection_zone_idx] =
+						RNSWS;
+				} else if (drv_data->compartment_owner[U(1)]) {
+					bkpreg_access[protection_zone_idx] =
+						RONS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			case BKPREG_ZONE_3_RIF1:
+				if (drv_data->compartment_owner[U(1)]) {
+					bkpreg_access[protection_zone_idx] =
+						RNSWNS;
+				} else if (drv_data->compartment_owner[U(0)] ||
+					   drv_data->compartment_owner[U(2)]) {
+					bkpreg_access[protection_zone_idx] =
+						RONS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			case BKPREG_ZONE_3_RIF0:
+				if (drv_data->compartment_owner[U(0)]) {
+					bkpreg_access[protection_zone_idx] =
+						RNSWNS;
+				} else if (drv_data->compartment_owner[U(1)] ||
+					   drv_data->compartment_owner[U(2)]) {
+					bkpreg_access[protection_zone_idx] =
+						RONS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			case BKPREG_ZONE_3_RIF2:
+				if (drv_data->compartment_owner[U(2)]) {
+					bkpreg_access[protection_zone_idx] =
+						RNSWNS;
+				} else if (drv_data->compartment_owner[U(0)] ||
+					   drv_data->compartment_owner[U(1)]) {
+					bkpreg_access[protection_zone_idx] =
+						RONS;
+				} else {
+					bkpreg_access[protection_zone_idx] =
+						NORNOW;
+				}
+				break;
+			default:
+				bkpreg_access = NULL;
+				break;
+			}
+
+			if (bkpreg_access == NULL) {
+				break;
+			}
+		}
+	}
+
+	return bkpreg_access;
+}
diff --git a/include/drivers/st/stm32mp_tamp_nvram.h b/include/drivers/st/stm32mp_tamp_nvram.h
new file mode 100644
index 0000000000..8c3d8ec165
--- /dev/null
+++ b/include/drivers/st/stm32mp_tamp_nvram.h
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef STM32MP_TAMP_NVRAM_H
+#define STM32MP_TAMP_NVRAM_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <drivers/st/nvmem.h>
+#include <dt-bindings/soc/rif.h>
+#include <lib/utils_def.h>
+
+#define TAMP_SECCFGR U(0x20)
+#define TAMP_BKPRIFR(x) (U(0x70) + (U(0x4) * (U(x) - U(1))))
+#define TAMP_RXCIDCFGR(x) (U(0x80) + (U(0x4) * (U(x))))
+#define CURRENT_CID U(RIF_CID1)
+
+#define REG_FIELD(_offset, _l, _h)                                 \
+	{                                                       \
+		.reg_offset = (_offset),				\
+		.lowest_bit = (_l),				\
+		.highest_bit = (_h)				\
+	}
+
+struct stm32_tamp_nvram_reg_field {
+	size_t reg_offset;
+	uint32_t lowest_bit;
+	uint32_t highest_bit;
+};
+
+enum stm32_tamp_bkpreg_access {
+	RONS,
+	RNSWNS,
+	RNSWS,
+	RSWS,
+	NORNOW,
+};
+
+struct stm32_tamp_nvram_drv_data {
+	uintptr_t base;
+	uintptr_t parent_base;
+	size_t size;
+	size_t parent_size;
+	const void *tamp_fdt;
+	int node;
+	size_t nb_total_regs;
+	int *bkpreg_zones_end;
+	enum stm32_tamp_bkpreg_access *bkpreg_access;
+	const struct stm32_tamp_nvram_cdata *cdata;
+	bool *compartment_owner;
+};
+
+struct stm32_tamp_nvram_cdata {
+	const unsigned int nb_zones;
+	const struct stm32_tamp_nvram_reg_field *zone_cfg;
+};
+
+//Implemented for MP1 and MP2
+
+enum stm32_tamp_bkpreg_access *
+stm32_tamp_nvram_get_access_rights(struct stm32_tamp_nvram_drv_data *drv_data);
+
+const struct stm32_tamp_nvram_cdata *stm32_tamp_nvram_get_cdata(void);
+
+
+//Shared in core
+int stm32_tamp_nvram_reg_field_write(uintptr_t base,
+				     struct stm32_tamp_nvram_reg_field field,
+				     uint32_t val);
+
+int stm32_tamp_nvram_reg_field_read(uintptr_t base,
+				    struct stm32_tamp_nvram_reg_field field,
+				    uint32_t *val);
+
+int stm32_tamp_nvram_reg_get_zone(struct stm32_tamp_nvram_drv_data *drv_data,
+				  int reg);
+
+#endif /* STM32MP_TAMP_NVRAM_H */
diff --git a/plat/st/common/common.mk b/plat/st/common/common.mk
index 4d1b0a2df3..740fe244b2 100644
--- a/plat/st/common/common.mk
+++ b/plat/st/common/common.mk
@@ -172,6 +172,8 @@ PLAT_BL_COMMON_SOURCES		+=	drivers/clk/clk.c				\
 					drivers/st/clk/stm32mp_clkfunc.c		\
 					drivers/st/ddr/stm32mp_ddr.c			\
 					drivers/st/gpio/stm32_gpio.c			\
+					drivers/st/nvmem/nvmem.c			\
+					drivers/st/nvmem/stm32mp_tamp_nvram_core.c	\
 					drivers/st/regulator/regulator_core.c		\
 					drivers/st/regulator/regulator_fixed.c		\
 					drivers/st/regulator/regulator_gpio.c			\
diff --git a/plat/st/common/include/stm32mp_common.h b/plat/st/common/include/stm32mp_common.h
index e83cacf9d6..9051336e42 100644
--- a/plat/st/common/include/stm32mp_common.h
+++ b/plat/st/common/include/stm32mp_common.h
@@ -174,6 +174,9 @@ uintptr_t stm32_get_bkpr_stop2_ep_addr(void);
 /* Display board information from the value found in OTP fuse */
 void stm32_display_board_info(uint32_t board_id);
 
+int stm32_tamp_nvram_init(void);
+int stm32_tamp_nvram_update_rights(void);
+
 #if PSA_FWU_SUPPORT
 uintptr_t stm32_get_bkpr_fwu_info_addr(void);
 void stm32_fwu_set_boot_idx(void);
diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index 6be99f18e2..1b9da0e8b3 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -319,6 +319,10 @@ void bl2_el3_plat_arch_setup(void)
 		panic();
 	}
 
+	if (stm32_tamp_nvram_init() < 0) {
+		panic();
+	}
+
 	stm32_save_boot_info(boot_context);
 
 #if STM32MP_USB_PROGRAMMER && STM32MP15
diff --git a/plat/st/stm32mp1/platform.mk b/plat/st/stm32mp1/platform.mk
index 04f03b8a07..8b9c135dc9 100644
--- a/plat/st/stm32mp1/platform.mk
+++ b/plat/st/stm32mp1/platform.mk
@@ -230,6 +230,8 @@ PLAT_BL_COMMON_SOURCES	+=	drivers/arm/tzc/tzc400.c				\
 				plat/st/stm32mp1/stm32mp1_helper.S			\
 				plat/st/stm32mp1/stm32mp1_syscfg.c
 
+PLAT_BL_COMMON_SOURCES  +=	drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
+
 ifeq ($(STM32MP13),1)
 PLAT_BL_COMMON_SOURCES	+=	drivers/st/clk/clk-stm32-core.c				\
 				drivers/st/clk/clk-stm32mp13.c				\
diff --git a/plat/st/stm32mp1/stm32mp1_def.h b/plat/st/stm32mp1/stm32mp1_def.h
index 4d3fb1dc68..6833cd6764 100644
--- a/plat/st/stm32mp1/stm32mp1_def.h
+++ b/plat/st/stm32mp1/stm32mp1_def.h
@@ -724,6 +724,7 @@ static inline uintptr_t tamp_bkpr(uint32_t idx)
 #define DT_RCC_SEC_CLK_COMPAT		"st,stm32mp1-rcc-secure"
 #endif
 #define DT_SDMMC2_COMPAT		"st,stm32-sdmmc2"
+#define DT_TAMP_NVRAM_COMPAT		"st,stm32mp15-tamp-nvram"
 #define DT_UART_COMPAT			"st,stm32h7-uart"
 
 #endif /* STM32MP1_DEF_H */
diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index b3eb003c2a..073f921b82 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -286,6 +286,10 @@ void bl2_el3_plat_arch_setup(void)
 		panic();
 	}
 
+	if (stm32_tamp_nvram_init() < 0) {
+		panic();
+	}
+
 	if (stm32_iwdg_init() < 0) {
 		panic();
 	}
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index fc2a0a1d6e..41590a0fd4 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -201,6 +201,8 @@ PLAT_BL_COMMON_SOURCES		+=	drivers/st/clk/clk-stm32-core.c				\
 					drivers/st/clk/stm32mp2_clk.c				\
 					drivers/st/crypto/stm32_rng.c
 
+PLAT_BL_COMMON_SOURCES  	+=	drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c
+
 BL2_SOURCES			+=	plat/st/stm32mp2/plat_bl2_mem_params_desc.c		\
 					plat/st/stm32mp2/stm32mp2_fconf_firewall.c
 
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 3f325769bb..9a28d16a64 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -575,6 +575,7 @@ static inline uintptr_t tamp_bkpr(uint32_t idx)
 #define DT_PWR_COMPAT				"st,stm32mp25-pwr"
 #define DT_RCC_CLK_COMPAT			"st,stm32mp25-rcc"
 #define DT_SDMMC2_COMPAT			"st,stm32mp25-sdmmc2"
+#define DT_TAMP_NVRAM_COMPAT			"st,stm32mp25-tamp-nvram"
 #define DT_UART_COMPAT				"st,stm32h7-uart"
 
 #endif /* STM32MP2_DEF_H */
-- 
2.53.0

